# In this file, we retest the Random Forest Regressor after bucketing features which had 0.8 or above collinearity.

In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

In [2]:
from pathlib import Path
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [3]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
clean_listing = pd.read_sql_query('select * from "clean_listing_remove_somereviews"',con=conn)
calendar = pd.read_sql_query('select * from "calendar_clean"',con=conn)
amenities = pd.read_sql_query('select * from "amenities_bucketed"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))

PostGres Download Duration: 2.836869955062866 seconds


In [4]:
# View clean_listing database
clean_listing.head()

id last_scraped host_since  host_listings_count host_is_superhost  \
0   6422   2019-01-17 2009-04-03                    1                 t   
1  20847   2019-01-17 2010-02-08                    2                 t   
2  25341   2019-01-17 2010-02-08                    2                 t   
3  25613   2019-01-17 2010-03-18                    1                 t   
4  37306   2019-01-17 2010-07-08                    1                 f   

  host_identity_verified neighbourhood_cleansed   latitude  longitude  \
0                      t             District 6  36.173151 -86.735813   
1                      f             District 7  36.196397 -86.704267   
2                      f             District 7  36.197053 -86.704937   
3                      t             District 6  36.178263 -86.741622   
4                      f            District 17  36.133109 -86.774767   

      room_type  ... cleaning_fee  review_scores_rating  number_of_reviews  \
0  Private room  ...          0.0                  99.0                561   
1  Private room  ...         25.0                  94.0                138   
2  Private room  ...         25.0                  96.0                102   
3  Private room  ...          0.0                  97.0                454   
4  Private room  ...          0.0                  99.0                100   

   instant_bookable is_business_travel_ready          cancellation_policy  \
0                 f                        f  strict_14_with_grace_period   
1                 f                        f  strict_14_with_grace_period   
2                 f                        f  strict_14_with_grace_period   
3                 t                        f                     moderate   
4                 f                        f  strict_14_with_grace_period   

   require_guest_profile_picture  require_guest_phone_verification  \
0                              f                                 f   
1                              t                                 t   
2                              t                                 t   
3                              t                                 t   
4                              t                                 t   

   has_availability  days_host  
0                 t       3576  
1                 t       3265  
2                 t       3265  
3                 t       3227  
4                 t       3115  

[5 rows x 27 columns]

In [5]:
# View data types of full_listings
clean_listing.dtypes

id                                           int64
last_scraped                        datetime64[ns]
host_since                          datetime64[ns]
host_listings_count                          int64
host_is_superhost                           object
host_identity_verified                      object
neighbourhood_cleansed                      object
latitude                                   float64
longitude                                  float64
room_type                                   object
property_type                               object
accommodates                                 int64
bathrooms                                  float64
bedrooms                                     int64
bed_type                                    object
price                                      float64
security_deposit                           float64
cleaning_fee                               float64
review_scores_rating                       float64
number_of_reviews              

In [6]:
# Drop duplicated variables.
listing = clean_listing.drop(columns = ['last_scraped', 'host_since', 'latitude', 'longitude', 'price'])

In [7]:
# View amenities dataframe
amenities.head()

id   tv  internet  wifi  air_conditioning  kitchen  \
0   6422  1.0       1.0   1.0               1.0      1.0   
1  20847  1.0       0.0   1.0               1.0      1.0   
2  25341  1.0       0.0   1.0               1.0      1.0   
3  25613  1.0       1.0   1.0               1.0      1.0   
4  37306  1.0       0.0   1.0               1.0      1.0   

   free_parking_on_premises  pets_live_on_this_property  dogs  cats  ...  \
0                       1.0                         1.0   1.0   1.0  ...   
1                       1.0                         0.0   0.0   0.0  ...   
2                       1.0                         0.0   0.0   0.0  ...   
3                       1.0                         0.0   0.0   0.0  ...   
4                       0.0                         1.0   1.0   1.0  ...   

   high-resolution_computer_monitor  standing_valet  pool_with_pool_hoist  \
0                               0.0             0.0                   0.0   
1                               0.0             0.0                   0.0   
2                               0.0             0.0                   0.0   
3                               0.0             0.0                   0.0   
4                               0.0             0.0                   0.0   

   sun_loungers  electric_profiling_bed  hammock  mobile_hoist  \
0           0.0                     0.0      0.0           0.0   
1           0.0                     0.0      0.0           0.0   
2           0.0                     0.0      0.0           0.0   
3           0.0                     0.0      0.0           0.0   
4           0.0                     0.0      0.0           0.0   

   Kitchen_Grouped_Binary  Bathroom_Grouped_Binary  Laundry_Grouped_Binary  
0                     1.0                      0.0                     1.0  
1                     0.0                      0.0                     1.0  
2                     1.0                      0.0                     1.0  
3                     0.0                      0.0                     1.0  
4                     0.0                      0.0                     1.0  

[5 rows x 164 columns]

In [8]:
# View calendar dataframe
calendar.head(10)

id      day  month  price
0  6422  weekday      1   40.0
1  6422  weekday      2   40.0
2  6422  weekday      3   40.0
3  6422  weekday      4   40.0
4  6422  weekday      5   40.0
5  6422  weekday      6   40.0
6  6422  weekday      7   40.0
7  6422  weekday      8   40.0
8  6422  weekday      9   40.0
9  6422  weekday     10   40.0

In [9]:
# Merge amenities table to full listings.
merge1 = listing.merge(amenities, how='left', on ='id')

In [10]:
# Merge calendar table to amenities and full listings.
merged = merge1.merge(calendar, how='left', on ='id')
merged.head(20)

id  host_listings_count host_is_superhost host_identity_verified  \
0   6422                    1                 t                      t   
1   6422                    1                 t                      t   
2   6422                    1                 t                      t   
3   6422                    1                 t                      t   
4   6422                    1                 t                      t   
5   6422                    1                 t                      t   
6   6422                    1                 t                      t   
7   6422                    1                 t                      t   
8   6422                    1                 t                      t   
9   6422                    1                 t                      t   
10  6422                    1                 t                      t   
11  6422                    1                 t                      t   
12  6422                    1                 t                      t   
13  6422                    1                 t                      t   
14  6422                    1                 t                      t   
15  6422                    1                 t                      t   
16  6422                    1                 t                      t   
17  6422                    1                 t                      t   
18  6422                    1                 t                      t   
19  6422                    1                 t                      t   

   neighbourhood_cleansed     room_type property_type  accommodates  \
0              District 6  Private room         House             2   
1              District 6  Private room         House             2   
2              District 6  Private room         House             2   
3              District 6  Private room         House             2   
4              District 6  Private room         House             2   
5              District 6  Private room         House             2   
6              District 6  Private room         House             2   
7              District 6  Private room         House             2   
8              District 6  Private room         House             2   
9              District 6  Private room         House             2   
10             District 6  Private room         House             2   
11             District 6  Private room         House             2   
12             District 6  Private room         House             2   
13             District 6  Private room         House             2   
14             District 6  Private room         House             2   
15             District 6  Private room         House             2   
16             District 6  Private room         House             2   
17             District 6  Private room         House             2   
18             District 6  Private room         House             2   
19             District 6  Private room         House             2   

    bathrooms  bedrooms  ... sun_loungers  electric_profiling_bed  hammock  \
0         1.0         2  ...          0.0                     0.0      0.0   
1         1.0         2  ...          0.0                     0.0      0.0   
2         1.0         2  ...          0.0                     0.0      0.0   
3         1.0         2  ...          0.0                     0.0      0.0   
4         1.0         2  ...          0.0                     0.0      0.0   
5         1.0         2  ...          0.0                     0.0      0.0   
6         1.0         2  ...          0.0                     0.0      0.0   
7         1.0         2  ...          0.0                     0.0      0.0   
8         1.0         2  ...          0.0                     0.0      0.0   
9         1.0         2  ...          0.0                     0.0      0.0   
10        1.0         2  ...          0.0                     0.0      0.0   
11        1.0         2

In [11]:
# Merge tables and drop 'id' (unique identifier - not relevant)
merged = merged.drop(columns = 'id')

In [12]:
# Create a copy of the database to work with going forward.
merged2 = merged.copy()

In [13]:
# Create a list of the object type columns.
objects = merged2.dtypes[merged2.dtypes == 'object'].index.tolist()
objects

['host_is_superhost',
 'host_identity_verified',
 'neighbourhood_cleansed',
 'room_type',
 'property_type',
 'bed_type',
 'instant_bookable',
 'is_business_travel_ready',
 'cancellation_policy',
 'require_guest_profile_picture',
 'require_guest_phone_verification',
 'has_availability',
 'day']

In [14]:
# Create a OneHotEncoder instance
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse=False, drop='if_binary')

# Fit and transform the OneHotEncoder using the categorical variable list
encode_df = pd.DataFrame(enc.fit_transform(merged2[objects]))

# Add the encoded variable names to the dataframe
encode_df.columns = enc.get_feature_names(objects)
encode_df.head()

host_is_superhost_t  host_identity_verified_t  \
0                  1.0                       1.0   
1                  1.0                       1.0   
2                  1.0                       1.0   
3                  1.0                       1.0   
4                  1.0                       1.0   

   neighbourhood_cleansed_District 1  neighbourhood_cleansed_District 10  \
0                                0.0                                 0.0   
1                                0.0                                 0.0   
2                                0.0                                 0.0   
3                                0.0                                 0.0   
4                                0.0                                 0.0   

   neighbourhood_cleansed_District 11  neighbourhood_cleansed_District 12  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 13  neighbourhood_cleansed_District 14  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 15  neighbourhood_cleansed_District 16  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   ...  cancellation_policy_flexible  cancellation_policy_moderate  \
0  ...                           0.0                           0.0   
1  ...                           0.0                           0.0   
2  ...                           0.0                           0.0   
3  ...                           0.0                           0.0   
4  ...                           0.0                           0.0   

   cancellation_policy_strict  \
0                         0.0   
1                         0.0   
2                         0.0   
3                         0.0   
4                         0.0   

   cancellation_policy_strict_14_with_grace_period  \
0                                              1.0   
1                                              1.0   
2                                              1.0   
3                                              1.0   
4                                              1.0   

   cancellation_policy_super_strict_30  cancellation_policy_super_strict_60  \
0                                  0.0                                  0.0   
1                                  0.0                                  0.0   
2                                  0.0                                  0.0   
3                                  0.0                                  0.0   
4                                  0.0                                  0.0   

   require_guest_profile_picture_t  require_guest_phone_verification_t  \
0                              0.0                                 0.0   
1                              0.0                                 0.0   
2                              0.0                                 0.0   
3                              0.0                                 0.0   
4                              0.0                    

In [15]:
# Merge one-hot encoded features and drop the originals
merged2 = merged2.merge(encode_df,left_index=True, right_index=True)
merged2 = merged2.drop(columns=objects)

In [16]:
merged2.head()

host_listings_count  accommodates  bathrooms  bedrooms  security_deposit  \
0                    1             2        1.0         2               0.0   
1                    1             2        1.0         2               0.0   
2                    1             2        1.0         2               0.0   
3                    1             2        1.0         2               0.0   
4                    1             2        1.0         2               0.0   

   cleaning_fee  review_scores_rating  number_of_reviews  days_host   tv  ...  \
0           0.0                  99.0                561       3576  1.0  ...   
1           0.0                  99.0                561       3576  1.0  ...   
2           0.0                  99.0                561       3576  1.0  ...   
3           0.0                  99.0                561       3576  1.0  ...   
4           0.0                  99.0                561       3576  1.0  ...   

   cancellation_policy_flexible  cancellation_policy_moderate  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

   cancellation_policy_strict  \
0                         0.0   
1                         0.0   
2                         0.0   
3                         0.0   
4                         0.0   

   cancellation_policy_strict_14_with_grace_period  \
0                                              1.0   
1                                              1.0   
2                                              1.0   
3                                              1.0   
4                                              1.0   

   cancellation_policy_super_strict_30  cancellation_policy_super_strict_60  \
0                                  0.0                                  0.0   
1                                  0.0                                  0.0   
2                                  0.0                                  0.0   
3                                  0.0                                  0.0   
4                                  0.0                                  0.0   

   require_guest_profile_picture_t  require_guest_phone_verification_t  \
0                              0.0                                 0.0   
1                              0.0                                 0.0   
2                              0.0                                 0.0   
3                              0.0                                 0.0   
4                              0.0                                 0.0   

   has_availability_t  day_weekend  
0                 1.0          0.0  
1                 1.0          0.0  
2                 1.0          0.0  
3                 1.0          0.0  
4                 1.0          0.0  

[5 rows x 259 columns]

In [17]:
pd.pandas.set_option('display.max_columns', None)

In [18]:
# Define the target variable.
y=merged2['price']

In [19]:
# Define the features.
X = merged2.drop(columns ='price')
# X = X.values.reshape(-1, 201)

In [20]:
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [21]:
X_test.shape

(35988, 258)

In [22]:
# import the regressor 
from sklearn.ensemble import RandomForestRegressor

#Instantiate the Decision Tree Regression model:
random_forest_regression_model = RandomForestRegressor(n_estimators = 100, random_state = 78) 
random_forest_regression_model.fit(X_train, y_train)

RandomForestRegressor(random_state=78)

In [23]:
# Calculate the r-squared score of the initial model.
R_squared_training_set = random_forest_regression_model.score(X_train, y_train, sample_weight=None)
print(f"R-Squared Training Set:  {R_squared_training_set}")

R_squared_testing_set = random_forest_regression_model.score(X_test, y_test, sample_weight=None)
print(f"R-Squared Testing Set:  {R_squared_testing_set}")

R-Squared Training Set:  0.9945417319609442
R-Squared Testing Set:  0.9721391831418923


In [24]:
print(merged2.shape)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(143952, 259)
(107964, 258)
(107964,)
(35988, 258)
(35988,)


In [25]:
# Now calculate the adjusted r-squared which adjusts the value by penalizing for excessive numbers of features
# Note, "k" is the number of predictors (features), n is the number of observations (rows)

#Formula:  Adjusted_R2 = 1 - (1-R2)*(n-1)/(n-k-1)

Adjusted_R_Squared_Training_Set = 1 - (1-R_squared_training_set)*(X_train.shape[0]-1)/(X_train.shape[0]-(273) - 1)
print(f"Adjusted R Squared for training set:  {Adjusted_R_Squared_Training_Set}")

Adjusted_R_Squared_Testing_Set = 1 - (1-R_squared_testing_set)*(X_test.shape[0]-1)/(X_test.shape[0]-(273) - 1)
print(f"Adjusted R Squared for testing set:  {Adjusted_R_Squared_Testing_Set}")

Adjusted R Squared for training set:  0.994527894954958
Adjusted R Squared for testing set:  0.9719262133540706


# From the above calculation using our testing set's r-squared value of 95.%, the adjusted r-sqaured value is nearly the same at 95.757%

# Now we will calculate the mean squared error and the mean error:

In [27]:
# Predict using the Decision Tree Random Forest Regression model.
y_pred = random_forest_regression_model.predict(X_test)
y_pred

array([455.95278091, 239.94125   , 464.807     , ..., 190.96444444,
       308.65366667,  46.97666667])

In [28]:
#Calculate the mean squared error
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
rmse = np.sqrt(mse)
print(f"Mean Error: {rmse}")

Mean Squared Error: 1889.5781915628966
Mean Error: 43.469278709945215


# It appears the collinearity bucketing adjustment allowed our Random Forest Regressor to predict our continuous target (price) with a mean squared error value of about 1889.58.  This indicates our model is off, on average, by about $43.47 (which is the square root of our mse)

# This is a slight improvement from the preliminary baseline test.